dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp

```diff
@@ -44,6 +47,43 @@ namespace DB
 {
 namespace PS::V3::tests
 {
+TEST(ExternalIdsByNamespace, Simple)
+{
+    NamespaceId ns_id = 100;
+    ExternalIdsByNamespace external_ids_by_ns;
+
+    std::atomic<Int32> who(0);
+
+    std::shared_ptr<PageIdV3Internal> holder = std::make_shared<PageIdV3Internal>(buildV3Id(ns_id, 50));
+
+    auto th_insert = std::async([&]() {
+        external_ids_by_ns.addExternalId(holder);
+
+        Int32 expect = 0;
+        who.compare_exchange_strong(expect, 1);
```

Comment:
What's the purpose of this line and the following line `who.compare_exchange_strong(expect, 2);`?

Reply:
Adding the `who` in the logging may help us find out the reason if a problem occurs unexpectly. It is a test to test whether there is any problem when `addExternalId` and `getAliveIds` running concurrency.

I tried to lock the `ns_id` instead of the whole map inside `ExternalIdsByNamespace`, but I found it is hard to make it bug-free: https://github.com/pingcap/tiflash/pull/5739/commits/2a1e2a351033af3be6a6ee5c94d69e90ec6a24e0#diff-71dc0bad723b5e11990be7e2a5d91ec338d333a16ee4a8bb780e65b69391c325R65-R71
So I simplify it to lock the whole map, which should be good enough. And this test is also simplified for just a notice.

Reply:
I got it. May be should use different `expect` value in these two lines and wait until `compare_exchange_strong` success?

Reply:
Got it.